File: tests/iris.py
import numpy as np;

import torch
import torch.nn as nn
import torch.nn.functional as F

# def kaiming_uniform_(
#     tensor: torch.Tensor,
#     a: float = 0,
#     mode: str = "fan_in",
#     nonlinearity: str = "leaky_relu"):
#     if 0 in tensor.shape:
#         torch.warnings.warn("Initializing zero-element tensors is a no-op")
#         return tensor
#     fan = torch.nn.init._calculate_correct_fan(tensor, mode)
#     gain = torch.nn.init.calculate_gain(nonlinearity, a)
#     std = gain / torch.math.sqrt(fan)
#     bound = torch.math.sqrt(3.0) * std  # Calculate uniform bounds from standard deviation
#     with torch.no_grad():
#         return tensor.uniform_(-bound, bound)

class Net(nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.id1 = nn.Identity()
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, num_classes, bias=False)
        self.id2 = nn.Identity()
        self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        out = self.fc1(x)
        out = self.id1(out)
        out = self.relu(out)
        out = self.fc2(out)
        out = self.id2(out)
        out = self.sigmoid(out)
        return out

    def get_num_params(self):
        n_params = sum(p.numel() for p in self.parameters())
        return n_params
    
    # def reset_parameters(self, linear) -> None:
    #     torch.manual_seed(137)
    #     print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    #     kaiming_uniform_(linear.weight, a=torch.math.sqrt(5))
    #     print(f"weight: {pretty_logits(linear.weight)}")
    #     print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    #     if linear.bias is not None:
    #         fan_in, _ = torch.nn.init._calculate_fan_in_and_fan_out(linear.weight)
    #         bound = 1 / torch.math.sqrt(fan_in) if fan_in > 0 else 0
    #         print(torch.randint(0, 0xFFFFFFFF, [1]).item())
    #         torch.nn.init.uniform_(linear.bias, -bound, bound)
    #         print(f"bias: {pretty_logits(linear.bias)}")
    #         print(torch.randint(0, 0xFFFFFFFF, [1]).item())

def pretty_logits(logits):
    logits0 = logits.view(-1)
    row = "["
    for j in range(min(len(logits0), 7)):
        row += f"{(round(logits0[j].item(), 4)):.4f}"
        if j == len(logits0) - 1:
            row += ""
        elif j > 0 and ((j + 1) % logits.shape[0]) == 0:
            row += ", "
        else:
            row += ", "
            
    row += "]"
    return row

def pretty_array(logits0):
    row = "["
    for j in range(min(len(logits0), 7)):
        row += f"{(round(logits0[j], 4)):.4f}"
        if j == len(logits0) - 1:
            row += ""
        else:
            row += ", "
    row += "]"
    return row

if __name__ == "__main__":
    import argparse

    from pathlib import Path

    path = Path(__file__).parent.absolute()

    parser = argparse.ArgumentParser()
    parser.add_argument("--batch_size", type=int, default=4)
    parser.add_argument("--optim", type=str, default="SGD")
    parser.add_argument("--lr", type=float, default=1e-3)
    parser.add_argument("--loss", type=str, default="MSELoss")
    args = parser.parse_args()

    assert 1 <= args.batch_size <= 1024
    assert args.optim in {"SGD", "AdamW"}
    assert 0 <= args.lr <= 1
    assert args.loss in {"BCELoss", "MSELoss"}

    data = np.loadtxt(str(path) + '//iris.csv', usecols=range(0,7), delimiter=",", skiprows=0, dtype=np.float32)

    def get_batch(B):
        assert B <= len(data), "not enough items for the specified batch size"
        i = 0
        while True:
            take = B
            if (i + take > len(data)):
                take = len(data) - i
            x = (data[i:i+take][:,[0, 1, 2, 3]])
            y = (data[i:i+take][:,[4, 5, 6]])
            yield x, y, take
            i += take
            if i >= len(data):
                i = 0

    # test data loader

    # data_iter_test = iter(get_batch(37))
    # for i in range(5):
    #     x, y, take = next(data_iter_test)
    #     print(f"batch_size: {take}:")
    #     for k in range(len(x)):
    #         print(f"x: {pretty_array(x[k])} = y: {pretty_array(y[k])}")

    torch.manual_seed(137)

    model = Net(input_size=4, hidden_size=8, num_classes=3)

    # test weight initialization

    print(f"fc1.weight: {pretty_logits(model.fc1.weight)}")
    if model.fc1.bias is not None:
        print(f"fc1.bias: {pretty_logits(model.fc1.bias)}")
    print(f"fc2.weight: {pretty_logits(model.fc2.weight)}")
    if model.fc2.bias is not None:
        print(f"fc2.bias: {pretty_logits(model.fc2.bias)}")

    if args.loss == "BCELoss":
        criterion = nn.BCELoss()
    if args.loss == "MSELoss":
        criterion = nn.MSELoss()

    if args.optim == "SGD":
        optimizer = torch.optim.SGD(model.parameters(), lr=args.lr)
    elif args.optim == "AdamW":
        optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr)

    print(f"parameters: {model.get_num_params()}")

    num_epochs = 10

    data_iter = iter(get_batch(args.batch_size))

    for epoch in range(num_epochs):
        x, y, take = next(data_iter)
        logits = model(torch.from_numpy(x))
        print(f'{epoch}: logits: {pretty_logits(logits)}')
        loss = criterion(logits, torch.from_numpy(y))
        print(f'{epoch}: loss: {loss.item():.4f}')
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        print(f"{epoch}: fc2.weight.grad: {pretty_logits(model.fc2.weight.grad)}")
        if model.fc2.bias is not None:
            print(f"{epoch}: fc2.bias.grad: {pretty_logits(model.fc2.bias.grad)}")
        print(f"{epoch}: fc1.weight.grad: {pretty_logits(model.fc1.weight.grad)}")
        if model.fc1.bias is not None:
            print(f"{epoch}: fc1.bias.grad: {pretty_logits(model.fc1.bias.grad)}")
        print(f"{epoch}: fc1.weight: {pretty_logits(model.fc1.weight)}")
        if model.fc1.bias is not None:
            print(f"{epoch}: fc1.bias: {pretty_logits(model.fc1.bias)}")
        print(f"{epoch}: fc2.weight: {pretty_logits(model.fc2.weight)}")
        if model.fc2.bias is not None:
            print(f"{epoch}: fc2.bias: {pretty_logits(model.fc2.bias)}")

File: tests/iris.cs
using System.Collections.Generic;
using System.Diagnostics;
using System;
using System.IO;
using System.Linq;

using nn;

unsafe internal static class iris {
    static IEnumerable<(float[][] x, float[][] y, uint B)> get_batch(string[] data, uint B) {
        int i = 0;
        while (true) {
            int take = checked((int)B);
            if (i + take > data.Length)
                take = data.Length - i;
            var x = data.Skip(i).Take(take).ToArray().Select(
                (s) => {
                    var split = s.Split(',').Take(4)
                        .Select(k => float.Parse(k)).ToArray();
                    return split;
                }).ToArray();
            var y = data.Skip(i).Take(take).ToArray().Select(
                (s) => {
                    var split = s.Split(',').Skip(4).Take(3)
                        .Select(k => float.Parse(k)).ToArray();
                    return split;
                }).ToArray();
            Debug.Assert(x.Length == take);
            Debug.Assert(y.Length == take);
            yield return (x, y, (uint)take);
            i += take;
            if (i >= data.Length)
                i = 0;
        }
    }

    static void reset_weights(Linear lin, IRNG g) {
        nn.init.kaiming_uniform_(
            lin._Weight.data,
            lin._Weight.numel(),
            g,
            lin.I,
            (float)Math.Sqrt(5));

        if (lin._Bias != null) {
            nn.rand.uniform_(
                lin._Bias.data,
                lin._Bias.numel(),
                g,
                -(float)(1.0 / Math.Sqrt(lin.I)),
                (float)(1.0 / Math.Sqrt(lin.I)));
        }
    }

    public static void run(TextWriter Console, string data_file, string optim, string loss_fn, float lr, uint batch_size) {

        var data = File.ReadAllLines(data_file);

        // test data loader batching

        // var data_iter_test = get_batch(data, 37).GetEnumerator();
        // data_iter_test.MoveNext();
        // for (int i = 0; i < 5; i++) {
        //     var c = data_iter_test.Current;
        //     Console.WriteLine($"batch_size: {c.len}:");
        //     for (int j = 0; j < c.len; j++) {
        //         Console.WriteLine($"x: [{string.Join(", ", c.x[j].Select(f => $"{f:f4}"))}] = y: [{string.Join(", ", c.y[j].Select(f => $"{f:f4}"))}]");
        //     }
        //     data_iter_test.MoveNext();
        // }

        var g = new nn.rand.mt19937(137);

        Linear fc1, fc2;

        nn.Sequential model = new nn.Sequential(
            fc1 = new Linear(4, 8, bias: true),
            new nn.Identity(),
            new nn.ReLU(),
            fc2 = new Linear(8, 3, bias: false),
            new nn.Identity(),
            new nn.Sigmoid()
        );

        reset_weights(fc1, g);
        reset_weights(fc2, g);

        Console.WriteLine($"fc1.weight: {Common.pretty_logits(fc1._Weight.data, fc1._Weight.numel())}");
        if (fc1._Bias != null) {
            Console.WriteLine($"fc1.bias: {Common.pretty_logits(fc1._Bias.data, fc1._Bias.numel())}");
        }
        Console.WriteLine($"fc2.weight: {Common.pretty_logits(fc2._Weight.data, fc2._Weight.numel())}");
        if (fc2._Bias != null) {
            Console.WriteLine($"fc2.bias: {Common.pretty_logits(fc2._Bias.data, fc2._Bias.numel())}");
        }

        var x = Tensor.zeros(batch_size * fc1.I);
        var y = new float[batch_size * fc2.O];

        for (int i = 0; i < y.Length; i++) {
            y[i] = float.NaN;
        }

        IOptimizer optimizer = null;

        if (optim == "SGD")
            optimizer = new nn.SGD(model.parameters(), lr: lr);
        else if (optim == "AdamW")
            optimizer = new nn.AdamW(model.parameters(), lr: lr);

        Console.WriteLine($"parameters: {optimizer.get_num_params()}");

        int epochs = 10;

        var data_iter = get_batch(data, batch_size).GetEnumerator();

        Console.WriteLine("train:");

        for (uint epoch = 0; epoch < epochs; epoch++) {
            data_iter.MoveNext();
            var sample = data_iter.Current;

            Console.WriteLine($"batch_size: {sample.B}");

            int n = 0;

            x.resize(sample.B * fc1.I);

            for (int i = 0; i < sample.B; i++) {
                for (int j = 0; j < sample.x[i].Length; j++) {
                    x.data[n++] = sample.x[i][j];
                }
            }

            Console.WriteLine($"x: {Common.pretty_logits(x.data, x.numel())}");

            n = 0;

            for (int i = 0; i < sample.B; i++) {
                for (int j = 0; j < sample.y[i].Length; j++) {
                    y[n++] = sample.y[i][j];
                }
            }

            var start_time = kernel32.millis();

            var logits = model.forward(x);

            Console.WriteLine($"{epoch}: logits: {Common.pretty_logits(logits.data, logits.numel())}");

            double loss = double.NaN;

            if (loss_fn == "BCELoss")
                loss = F.binary_cross_entropy(
                    logits,
                    y);
            else if (loss_fn == "MSELoss")
                loss = F.mse_loss(
                    logits.data,
                    logits.grad,
                    logits.numel(),
                    y);

            Console.WriteLine($"{epoch}: loss: {loss:f4}");

            optimizer.zero_grad();

            model.backward(logits);

            var elapsedMillis = (kernel32.millis() - start_time);

            optimizer.step();

            Console.WriteLine($"{epoch}: fc2.weight.grad: {Common.pretty_logits(fc2._Weight.grad, fc2._Weight.numel())}");
            if (fc2._Bias != null)
                Console.WriteLine($"{epoch}: fc2.bias.grad: {Common.pretty_logits(fc2._Bias.grad, fc2._Bias.numel())}");
            Console.WriteLine($"{epoch}: fc1.weight.grad: {Common.pretty_logits(fc1._Weight.grad, fc1._Weight.numel())}");
            if (fc1._Bias != null)
                Console.WriteLine($"{epoch}: fc1.bias.grad: {Common.pretty_logits(fc1._Bias.grad, fc1._Bias.numel())}");

            Console.WriteLine($"{epoch}: fc1.weight: {Common.pretty_logits(fc1._Weight.data, fc1._Weight.numel())}");
            if (fc1._Bias != null)
                Console.WriteLine($"{epoch}: fc1.bias: {Common.pretty_logits(fc1._Bias.data, fc1._Bias.numel())}");
            Console.WriteLine($"{epoch}: fc2.weight: {Common.pretty_logits(fc2._Weight.data, fc2._Weight.numel())}");
            if (fc2._Bias != null)
                Console.WriteLine($"{epoch}: fc2.bias: {Common.pretty_logits(fc2._Bias.data, fc2._Bias.numel())}");
        }

        data_iter = get_batch(data, 1).GetEnumerator();

        Console.WriteLine("eval:");

        x.resize(fc1.I);

        for (uint s = 0; s < data.Length; s++) {
            data_iter.MoveNext();
            var sample = data_iter.Current;
            int n = 0;
            Debug.Assert(sample.B == 1);
            for (int j = 0; j < sample.x[0].Length; j++) {
                x.data[n++] = sample.x[0][j];
            }
            var logits = model.forward(x);
            Console.WriteLine($"{s}: logits: {Common.pretty_logits(logits.data, logits.numel())}");
        }
    }
}